File: my-folder/0127-word-ladder/solution.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> set = new HashSet<>(wordList);
        Set<String> visited = new HashSet<>();
        Queue<String> q = new LinkedList<>();
        q.offer(beginWord);
        visited.add(beginWord);
        // curString -> parent string
        Map<String, String> map = new HashMap<>();
        
        int len = 1;
        while(!q.isEmpty()){
            int size = q.size();
            for(int i=0;i<size;i++){
                String word = q.poll();
                if(word.equals(endWord)){
                    // Deque<String> stack = new ArrayDeque<>();
                    // stack.push(word);
                    // String parent = map.get(word);
                    // while(!parent.equals(beginWord)){
                    //     stack.push(parent);
                    //     parent = map.get(parent);                
                    // }
                    // stack.push(parent);
                    // while(!stack.isEmpty()){
                    //     System.out.println(stack.pop());
                    // }
                   
                    return len;
                }
                char[] chars = word.toCharArray();
                
                for(int j=0;j<chars.length;j++){
                    char old = chars[j];
                    for(int c='a';c<='z';c++){
                        chars[j] = (char)c;
                        String newWord = new String(chars);
                        if(set.contains(newWord)&&!visited.contains(newWord)){
                            q.offer(newWord);
                            visited.add(newWord);
                            map.put(newWord, word);
                        }
                        chars[j] = old;
                    }
                }
            }
            ++len;
        }
        return 0;
        
    }
}

File: my-folder/0127-word-ladder/solution.py
class Solution(object):
    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        # if beginWord==endWord:
        #     return 1
        
        word_dict = collections.defaultdict(set)
        for word in wordList:
            for i in range(len(word)):
                # if i<len(word):
                    word_dict[word[:i]+"_"+word[i+1:]].add(word)
                # else:
                #     word_dict[word[:i]+"_"].add(word)
        
        # step = 0
        
        # def bfs(now_word):
        #     for i in range(len(word)):
        #         if i<len(word)-1:
        #             temp_word = word[:i]+"_"+word[i+1:]
        #         else:
        #             temp_word = word[:i]+"_"
        #     for
        step = 0
        q = [beginWord]
        visited = set([beginWord])
        while q:
            sz = len(q)
            step+=1
            for _ in range(sz):
                word = q.pop(0)
                if word==endWord:
                    return step
                for i in range(len(word)):
                    # if i<len(word)-1:
                    temp_word = word[:i]+"_"+word[i+1:]
                    # else:
                    #     temp_word = word[:i]+"_"
                    for nxt in word_dict[temp_word]:
                        # if nxt==endWord:
                        #     return step
                        # else:
                            if nxt not in visited:
                                visited.add(nxt)
                                q.append(nxt)
        return 0

                
        
